File: Src/ExpectationLib/ContactCircumstance.cs
namespace ExpectationLib;

/// <summary>
/// A particular response to a contact stimulus 
/// </summary>
public class ContactCircumstance : ICircumstance
{
    /// <inheritdoc />
    public IStimulus Stimulus { get; }
    
    /// <inheritdoc />
    public IResponse Response => new ContactResponse(_responseContext, From);
    
    private IParty From { get; }
    private readonly string _responseContext;

    public ContactCircumstance(IStimulus stimulus, string responseContext, IParty overrideFrom = null)
    {
        Stimulus = stimulus;
        From = overrideFrom ?? Stimulus.Receiver;
        _responseContext = responseContext;
    }
}

File: Src/CppLibExpectation/ContactCircumstance.cpp
#include "ContactCircumstance.h"
#include <sstream>
#include "ContactResponse.h"
#include "ContactStimulus.h"


namespace ExpectationLib
{
	ContactCircumstance::ContactCircumstance(const std::shared_ptr<IResponse>& response)
	{
		this->from = response->GetReceiver();
		this->response = response;
		this->stimulus = response->GetStimulus();

		std::stringstream circumstanceId;
		circumstanceId << GetResponse()->GetSender()->GetId()
					   << "-[" << GetResponse()->GetContext() << "]->"
					   << GetResponse()->GetReceiver()->GetId();
		this->id = circumstanceId.str();
	}

	std::shared_ptr<IStimulus> ContactCircumstance::GetStimulus()
	{
		return response->GetStimulus();
	}

	std::shared_ptr<IResponse> ContactCircumstance::GetResponse()
	{
		return response;
	}

	std::shared_ptr<IParty> ContactCircumstance::GetFrom()
	{
		return from;
	}

	const std::string ContactCircumstance::GetId()
	{		
		return id;
	}
}
